core/commands/storage.go

```diff
@@ -245,19 +260,34 @@ the chunk and replies back to client for the next challenge step.`,
 			return fmt.Errorf("fail to get peer ID from request")
 		}
 
+		// build session
+		ss := storage.SessionMap[ssID]
+		if ss == nil {
```

Comment:
If this is nil, it should *not* create. Session must be created in `/upload` for each chunk, not when `/upload/init` is called (because you already passed in `session-id`!)

Reply:
Yes, but when a host first starts, it won't have any session info, when a client connects to it, it will get client's sessionID and create its own copy here. I have to new session but reused client ssID

Reply:
Oh, I see, makes more sense now

Reply:
I see what the confusion is now.. `NewSession` should not be performed on a single `*session`, but it should be part of the `SessionMap`'s mutex...

Reply:
So you mean I shouldn't pass any ssID in new Session to put in SessionMap? Then where can I use lock to put into a map? an Extra mutex to do that in session.go?